crates/red_knot/src/db.rs

```diff
@@ -0,0 +1,327 @@
+    fn enter_node(&mut self, node: AnyNodeRef) -> TraversalSignal {
+        // Don't traverse into expressions
+        if node.is_expression() {
+            return TraversalSignal::Skip;
+        }
+
+        TraversalSignal::Traverse
+    }
+
+    fn visit_stmt(&mut self, stmt: &Stmt) {
+        match stmt {
+            Stmt::Import(import) => {
+                for alias in &import.names {
+                    let path: PathBuf = alias.name.split('.').collect();
+
+                    self.push_dependency(path.with_extension("py"));
+                }
+            }
+
+            Stmt::ImportFrom(from) => {
+                if let Some(module) = &from.module {
+                    let path: PathBuf = module.split('.').collect();
+                    self.push_dependency(path.with_extension("py"));
+                } else if let Some(level) = &from.level {
+                    let path: PathBuf = (0..*level).map(|_| "..").collect();
+                    self.push_dependency(path.with_extension("py"));
+                } else {
+                    // Should never happen, let's assume we didn't see it.
+                }
+            }
+            _ => {}
+        }
+        preorder::walk_stmt(self, stmt);
+    }
+}
+
+#[derive(Eq, PartialEq, Hash, Clone, Debug)]
+pub struct Dependency {
+    // A relative path from the current module to the dependency
+    pub path: PathBuf,
+}
+
+// TODO it's unclear to me if the function should accept a parsed or a source text?
+//   Is it best practice to inline as many db calls or should we ask the caller to do the db calls?
+#[salsa::tracked(jar=SourceJar)]
+pub fn check_syntax(db: &dyn Db, parsed: Parsed) -> SyntaxCheck {
+    // TODO I haven't looked into how many rules are pure syntax checks.
+    //   It may be necessary to at least give access to a simplified semantic model.
+    struct SyntaxChecker {
+        diagnostics: Vec<String>,
+    }
+
+    impl Visitor<'_> for SyntaxChecker {
+        fn visit_expr(&mut self, expr: &'_ Expr) {
+            if let Expr::Name(name) = expr {
+                if &name.id == "a" {
+                    self.diagnostics.push("Use of name a".to_string());
+                }
+            }
+        }
+    }
+
+    let mut visitor = SyntaxChecker {
+        diagnostics: Vec::new(),
+    };
+    visitor.visit_body(&parsed.ast(db).body);
+
+    SyntaxCheck::new(db, visitor.diagnostics)
+}
+
+#[salsa::tracked(jar=SourceJar)]
+pub fn check_physical_lines(db: &dyn Db, source_text: SourceText) -> PhysicalLinesCheck {
+    let text = source_text.text(db);
+
+    let mut diagnostics = Vec::new();
+    let mut line_number = 0u32;
+    for line in text.lines() {
+        if line.chars().count() > 88 {
+            diagnostics.push(format!("Line {} too long", line_number + 1));
+        }
+        line_number += 1;
+    }
+
+    PhysicalLinesCheck::new(db, diagnostics)
+}
+
+#[derive(Eq, PartialEq, Hash, Debug)]
+pub struct HirAstId<N: HasAstId> {
```

Comment:
Any reason this is defined in `db.rs` rather than either in `ast_id.rs` or `hir.rs`?

Reply:
No, there's none. I should probably have moved it when I introduced `hir.rs`